### ISSUE TITLE
Incorrect kernel parameters with sparse models

### ISSUE THREAD

**Issue by ctorney** [GPflow/GPflow#883]
I've noticed the following behaviour using sparse GPs that I'd like to understand better. I generate some data to fit as follows

```
kern=gpflow.kernels.Matern12(1,lengthscales=5.0,variance=1.0)
N=500
X = np.linspace(0, N, N)[:,None]
K = kern.compute_K_symm(X)

Y = np.random.multivariate_normal(np.zeros(N), K, 1).T
Y += (0.1)*np.random.normal(size=Y.shape)
```

As a sanity check I use a sparse GP with half the data and the same number of inducing points
```
sgpr = gpflow.models.SGPR(X[::2], Y[::2], kern=gpflow.kernels.Matern12(1,variance=1,lengthscales=5), Z=X[::2].copy())
sgpr.likelihood.variance = 0.01

gpflow.train.ScipyOptimizer().minimize(sgpr)
sgpr.as_pandas_table()
```
As expected this gives the same results as GPR and I recover the correct parameters:
> SGPR/feature/Z 	Parameter 	None 	(none) 	True 	(250, 1) 	True 	[[-3.1508541606040293e-06], [2.004008026951309...
> SGPR/kern/lengthscales 	Parameter 	None 	+ve 	True 	() 	True 	5.008733368270208
> SGPR/kern/variance 	Parameter 	None 	+ve 	True 	() 	True 	0.9601464571035354
> SGPR/likelihood/variance 	Parameter 	None 	+ve 	True 	() 	True 	0.010013976529520336
> 

Next using all the data and half as many inducing points
```
sgpr = gpflow.models.SGPR(X, Y, kern=gpflow.kernels.Matern12(1,variance=1,lengthscales=5), Z=X[::2].copy())
sgpr.likelihood.variance = 0.01

gpflow.train.ScipyOptimizer().minimize(sgpr,maxiter=2000)
sgpr.as_pandas_table()
```

this gives

> SGPR/feature/Z 	Parameter 	None 	(none) 	True 	(250, 1) 	True 	[[2.499022449377244e-06], [2.004008178124323],...
> SGPR/kern/lengthscales 	Parameter 	None 	+ve 	True 	() 	True 	14.086365818290883
> SGPR/kern/variance 	Parameter 	None 	+ve 	True 	() 	True 	0.8030125251127183
> SGPR/likelihood/variance 	Parameter 	None 	+ve 	True 	() 	True 	0.2255555730664038

So even though I initialised with the known parameters the optimisation moves away. I've tried this with other optimizers (Adam, Adagrad) and with SVGP using natural gradients and get the same thing. I made a plot of the objective function value and the kernel lengthscale during training with Adam for the first case using half the data and the same number of inducing points here:
![half_data_half_inducing_points](https://user-images.githubusercontent.com/3384313/48941650-589f0100-ef14-11e8-81cb-297ab4566db8.png)
and the second case where the lengthscale keeps increasing here:
![all_data_less_inducing_points](https://user-images.githubusercontent.com/3384313/48941683-75d3cf80-ef14-11e8-90fa-b4a99da81f8f.png)

But if I use an RBF kernel I get the correct behaviour in both cases. Is this a numerical stability issue, and there's something about the roughness of the matern12 kernel that leads to this behaviour? Or maybe I'm implementing the sparse GPs incorrectly. Any advice would be appreciated.
I'm using gpflow 1.3, tf 1.10.1

**Comment by hughsalimbeni** [comment]
This is to be expected. The problem is that slack in the bound is not uniform over hyperparameter settings. There is no reason that argmax(ELBO) is the same as argmax(true marginal likelihood), where the argmax is wrt model hyperparameters. If the ELBO is used as a proxy for the marginal likelihood then models will be learned that make the sparse approximation better, e.g. long lengthscales.

**Comment by ctorney** [comment]
thanks Hugh - so in general are you saying the sparse methods can't be used to learn model hyperparameters? i.e. only learn hyperparameters on a subset of the data with a full gp?

**Comment by hughsalimbeni** [comment]
Sparse methods certainly can be used to learn hyperparameters, it's just that there's no guarantee that they will be the same hyperparameters as the non-sparse model. The bias introduced through optimizing the ELBO rather than the ML will favour models that are well approximated by the sparse posterior, which is likely to be a good thing if computation constraints force the use of a sparse model in the first place. NB subset of data is likely to be worse.
